Feature/nathantypanski/commands

nathantypanski opened the pull request:
Everything up until "echo". Basic string library.

nathantypanski committed:
allocator-debugging basic string implementation

nathantypanski committed:
use asm clobbers

nathantypanski committed:
better putnum(), move memory offset again

nathantypanski committed:
closes #4: remove debug printing

nathantypanski committed:
rename send_at_c to "map"

nathantypanski committed:
align at a multiple of 8

nathantypanski committed:
make the prompt a static string

nathantypanski committed:
remove memory debug stuff

nathantypanski committed:
add "history" command

nathantypanski committed:
part of #5: implement echo

nathantypanski committed:
remove old kernel code; cleanup equality

Also remove the useless extra alloc in cstr::news(). That was leftover
from testing the allocator.

The equality functions are rewritten to make use of get_char().

nathantypanski committed:
Revert "remove old kernel code; cleanup equality"

This reverts commit 3711d2aa7aac214e86344165828102906664ec13.

It broke string equality. A rewrite still needs to happen, but this
isn't the one we should use.

nathantypanski committed:
kill getarg()

nathantypanski committed:
only push to history if command is nonempty

nathantypanski committed:
kill "history" command; simplify cstr impl

lenary commented on kernel/cstr.rs:
```diff
@@ -0,0 +1,230 @@
+        }
+    }
+    
+    #[allow(dead_code)]
```
do we get a warning if we get rid of this declaration?

lenary commented on kernel/memory/allocator.rs:
```diff
@@ -194,10 +195,13 @@ impl Allocator for Alloc {
+            //sgash::putchar('o');
+            //sgash::putnum(roffset as uint, 0);
+            //sgash::putchar('s');
+            //sgash::putnum(rsize as uint, 0);
```
these really are dead bits of code. Just delete them :)

lenary commented:
Address those comments then I'll merge (I'll probably [squash](http://git-scm.com/book/ch6-4.html#Squashing-Commits) in the merge to have fewer commits in the history, but that shouldn't be an issue)

nathantypanski committed:
fix dead code and unused warnings

nathantypanski commented:
Fixed in 998528d.

nathantypanski committed:
closes #5: recognize the rest of the commands

This adds command recognition for the following:

"ls, cat, cd, rm, mkdir, pwd, wr"

The command "echo" was already implemented.

nathantypanski committed:
don't count partial string matches as equal

Previously, typing e.g. "cda" at the prompt would count as a match for
"cd", since streq() wasn't checking to ensure the length of the cstr
wasn't longer than that of the &str (it was, however, checking if it was
shorter).

This solves that issue by comparing the length at the end of the
equality test.

lenary commented:
This all looks good. I'll hit merge.

lenary commented:
in fact, merge when you're ready.
